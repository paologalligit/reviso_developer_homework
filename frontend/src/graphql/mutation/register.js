import gql from 'graphql-tag';

export default gql`
  mutation(
    $name: String!
    $surname: String!
    $username: String!
    $email: String!
    $password: String!
    $birthdate: String!
    $country: String!
    $city: String!
    $address: String!
    $postal: Int!
  ) {
    registerUser(
      name: $name
      surname: $surname
      username: $username
      email: $email
      password: $password
      birthdate: $birthdate
      country: $country
      city: $city
      address: $address
      postal: $postal
    ) {
      ok
      errors {
        path
        message
      }
    }
  }
`;
