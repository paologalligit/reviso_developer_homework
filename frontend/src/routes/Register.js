/* eslint-disable no-restricted-syntax */
/* eslint-disable guard-for-in */
/* eslint-disable react/prop-types */
/* eslint-disable react/destructuring-assignment */
import React from 'react';
import {
  Message, Button, Input, Container, Header, Form, Grid,
} from 'semantic-ui-react';
import { graphql } from 'react-apollo';

import style from './styles/RegisterStyle';
import registerMutation from '../graphql/mutation/register';

class Register extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name: '',
      surname: '',
      username: '',
      email: '',
      password: '',
      birthdate: '',
      country: '',
      city: '',
      address: '',
      postal: '',
      errors: {},
    };
  }

  onSubmit = async () => {
    const {
      name,
      surname,
      username,
      email,
      password,
      birthdate,
      country,
      city,
      address,
      postal,
    } = this.state;

    let intPostal;

    if (postal === '') {
      intPostal = -1;
    } else {
      intPostal = parseInt(postal, 10);
    }

    const response = await this.props.mutate({
      variables: {
        name,
        surname,
        username,
        email,
        password,
        birthdate,
        country,
        city,
        address,
        postal: intPostal,
      },
    });


    const { ok, errors } = response.data.registerUser;

    if (ok) {
      this.props.history.push('/');
    } else {
      // console.log('the errors, smthg went wrong: ', errors);
      const err = {};
      errors.forEach(({ path, message }) => {
        err[`${path}Error`] = message;
      });

      this.setState({ errors: err });
    }
  };

  onChange = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  };

  render() {
    const {
      name,
      surname,
      username,
      email,
      password,
      birthdate,
      country,
      city,
      address,
      postal,
      errors,
    } = this.state;

    const {
      nameError,
      surnameError,
      usernameError,
      emailError,
      passwordError,
      birthdateError,
      countryError,
      cityError,
      addressError,
      postalError,
    } = errors;

    const errorsList = new Set();
    for (const child in errors) {
      errorsList.add(errors[child]);
    }

    return (
      <Container text textAlign="center" style={style.register}>
        <Header as="h2">Register</Header>
        <Form>
          <Grid>
            <Grid.Row columns={3}>
              <Grid.Column>
                <Form.Field error={!!nameError}>
                  <Input
                    name="name"
                    onChange={this.onChange}
                    value={name}
                    placeholder="Name"
                    fluid
                  />
                </Form.Field>
              </Grid.Column>
              <Grid.Column>
                <Form.Field error={!!surnameError}>
                  <Input
                    name="surname"
                    onChange={this.onChange}
                    value={surname}
                    placeholder="Surname"
                    fluid
                  />
                </Form.Field>
              </Grid.Column>
              <Grid.Column>
                <Form.Field error={!!usernameError}>
                  <Input
                    name="username"
                    onChange={this.onChange}
                    value={username}
                    placeholder="Username"
                    fluid
                  />
                </Form.Field>
              </Grid.Column>
            </Grid.Row>
            <Grid.Row columns={3}>
              <Grid.Column>
                <Form.Field error={!!emailError}>
                  <Input
                    name="email"
                    onChange={this.onChange}
                    value={email}
                    placeholder="Email"
                    fluid
                  />
                </Form.Field>
              </Grid.Column>
              <Grid.Column>
                <Form.Field error={!!passwordError}>
                  <Input
                    name="password"
                    onChange={this.onChange}
                    value={password}
                    type="password"
                    placeholder="Password"
                    fluid
                  />
                </Form.Field>
              </Grid.Column>
              <Grid.Column>
                <Form.Field error={!!birthdateError}>
                  <Input
                    name="birthdate"
                    onChange={this.onChange}
                    value={birthdate}
                    placeholder="Birthdate"
                    fluid
                  />
                </Form.Field>
              </Grid.Column>
            </Grid.Row>

            <Grid.Row columns={3}>
              <Grid.Column>
                <Form.Field error={!!countryError}>
                  <Input
                    name="country"
                    onChange={this.onChange}
                    value={country}
                    placeholder="Country"
                    fluid
                  />
                </Form.Field>
              </Grid.Column>
              <Grid.Column>
                <Form.Field error={!!cityError}>
                  <Input
                    name="city"
                    onChange={this.onChange}
                    value={city}
                    placeholder="City"
                    fluid
                  />
                </Form.Field>
              </Grid.Column>
              <Grid.Column>
                <Form.Field error={!!addressError}>
                  <Input
                    name="address"
                    onChange={this.onChange}
                    value={address}
                    placeholder="Address"
                    fluid
                  />
                </Form.Field>
              </Grid.Column>
            </Grid.Row>

            <Grid.Row columns={1}>
              <Grid.Column>
                <Form.Field error={!!postalError}>
                  <Input
                    name="postal"
                    onChange={this.onChange}
                    value={postal}
                    placeholder="Postal"
                    fluid
                  />
                </Form.Field>
              </Grid.Column>
            </Grid.Row>
            <Grid.Row columns={1}>
              <Grid.Column textAlign="center">
                <Button onClick={this.onSubmit} size="huge" primary>
                  Submit
                </Button>
              </Grid.Column>
            </Grid.Row>
          </Grid>
        </Form>
        {errorsList.size ? (
          <Message
            error
            header="There was some errors with your submission"
            list={Array.from(errorsList)}
          />
        ) : null}
      </Container>
    );
  }
}

export default graphql(registerMutation)(Register);
