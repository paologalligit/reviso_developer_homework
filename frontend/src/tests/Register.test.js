/* eslint-disable func-names */
import 'jsdom-global/register';
import React from 'react';
import toJson from 'enzyme-to-json';
import { mount, render } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

import registerMutation from '../graphql/mutation/register';
import Register from '../routes/Register';

const waitForData = () => new Promise(res => setTimeout(res, 100));

describe('register renders', () => {
  const mocks = [
    {
      request: {
        query: registerMutation,
        variables: {
          name: 'paolo',
          surname: 'galli',
          username: 'pgalli',
          email: 'p.galli@email.com',
          password: 'bobobo',
          birthdate: '1991-12-31',
          country: 'Italy',
          city: 'Milan',
          address: 'Via ciancihi, 98',
          postal: 20100,
        },
      },
      result: {
        data: {
          ok: true,
          user: {
            id: 1,
          },
          errors: null,
        },
      },
    },
  ];
  let wrapper;

  // initialization
  beforeEach(() => {
    wrapper = render(
      <MockedProvider mocks={mocks}>
        <Register />
      </MockedProvider>,
    );
  });

  it('matches snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders login page', () => {
    expect(wrapper).toHaveLength(1);
  });
});

describe('register interaction', () => {
  const correctMockQuery = [
    {
      request: {
        query: registerMutation,
        variables: {
          name: 'paolo',
          surname: 'galli',
          username: 'pgalli',
          email: 'p.galli@email.com',
          password: 'bobobo',
          birthdate: '1991-12-31',
          country: 'Italy',
          city: 'Milan',
          address: 'Via ciancichi, 98',
          postal: 20100,
        },
      },
      result: {
        data: {
          registerUser: {
            ok: true,
            errors: null,
          },
        },
      },
    },
  ];
  const wrongMockQuery = [
    {
      request: {
        query: registerMutation,
        variables: {
          name: 'paolo',
          surname: 'galli',
          username: 'pgalli',
          email: 'p.galli@email.com',
          password: 'bobobo',
          birthdate: '1991-12-31',
          country: 'Italy',
          city: 'Milan',
          address: 'Via ciancichi, 98',
          postal: 20100,
        },
      },
      result: {
        data: {
          registerUser: {
            ok: false,
            errors: [
              {
                path: 'username',
                message: 'username must be unique',
              },
            ],
          },
        },
      },
    },
  ];
  const historyMock = { push: jest.fn(path => path) };

  it('navigate to / with correct login', async () => {
    const wrapper = mount(
      <MockedProvider mocks={correctMockQuery} addTypename={false}>
        <Register history={historyMock} />
      </MockedProvider>,
    );
    expect(wrapper).toHaveLength(1);
    const inputs = wrapper.find('Input');
    const name = inputs.at(0).find('input');
    const surname = inputs.at(1).find('input');
    const username = inputs.at(2).find('input');
    const email = inputs.at(3).find('input');
    const password = inputs.at(4).find('input');
    const birthdate = inputs.at(5).find('input');
    const country = inputs.at(6).find('input');
    const city = inputs.at(7).find('input');
    const address = inputs.at(8).find('input');
    const postal = inputs.at(9).find('input');
    const button = wrapper.find('Button');

    name.simulate('change', { target: { name: 'name', value: 'paolo' } });
    surname.simulate('change', { target: { name: 'surname', value: 'galli' } });
    username.simulate('change', { target: { name: 'username', value: 'pgalli' } });
    email.simulate('change', { target: { name: 'email', value: 'p.galli@email.com' } });
    password.simulate('change', { target: { name: 'password', value: 'bobobo' } });
    birthdate.simulate('change', { target: { name: 'birthdate', value: '1991-12-31' } });
    country.simulate('change', { target: { name: 'country', value: 'Italy' } });
    city.simulate('change', { target: { name: 'city', value: 'Milan' } });
    address.simulate('change', { target: { name: 'address', value: 'Via ciancichi, 98' } });
    postal.simulate('change', { target: { name: 'postal', value: 20100 } });

    button.simulate('click');

    await waitForData();

    expect(historyMock.push).toHaveBeenCalledTimes(1);
    expect(historyMock.push.mock.calls[0]).toEqual(['/']);
  });

  it('login errors with wrong fields', async () => {
    const wrapper = mount(
      <MockedProvider mocks={wrongMockQuery} addTypename={false}>
        <Register history={historyMock} />
      </MockedProvider>,
    );
    expect(wrapper).toHaveLength(1);
    const inputs = wrapper.find('Input');
    const name = inputs.at(0).find('input');
    const surname = inputs.at(1).find('input');
    const username = inputs.at(2).find('input');
    const email = inputs.at(3).find('input');
    const password = inputs.at(4).find('input');
    const birthdate = inputs.at(5).find('input');
    const country = inputs.at(6).find('input');
    const city = inputs.at(7).find('input');
    const address = inputs.at(8).find('input');
    const postal = inputs.at(9).find('input');
    const button = wrapper.find('Button');

    name.simulate('change', { target: { name: 'name', value: 'paolo' } });
    surname.simulate('change', { target: { name: 'surname', value: 'galli' } });
    username.simulate('change', { target: { name: 'username', value: 'pgalli' } });
    email.simulate('change', { target: { name: 'email', value: 'p.galli@email.com' } });
    password.simulate('change', { target: { name: 'password', value: 'bobobo' } });
    birthdate.simulate('change', { target: { name: 'birthdate', value: '1991-12-31' } });
    country.simulate('change', { target: { name: 'country', value: 'Italy' } });
    city.simulate('change', { target: { name: 'city', value: 'Milan' } });
    address.simulate('change', { target: { name: 'address', value: 'Via ciancichi, 98' } });
    postal.simulate('change', { target: { name: 'postal', value: 20100 } });

    button.simulate('click');

    await waitForData();

    const loginState = wrapper.find('Register').instance().state;
    expect(historyMock.push).toHaveBeenCalledTimes(0);
    expect(loginState).toEqual({
      address: 'Via ciancichi, 98',
      birthdate: '1991-12-31',
      city: 'Milan',
      country: 'Italy',
      email: 'p.galli@email.com',
      errors: { usernameError: 'username must be unique' },
      name: 'paolo',
      password: 'bobobo',
      postal: 20100,
      surname: 'galli',
      username: 'pgalli',
    });
  });
});
