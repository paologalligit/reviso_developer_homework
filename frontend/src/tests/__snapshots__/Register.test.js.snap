// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register renders matches snapshot 1`] = `
<div
  class="ui text center aligned container"
  style="margin-top:50px;margin-bottom:50px"
>
  <h2
    class="ui header"
  >
    Register
  </h2>
  <form
    class="ui form"
  >
    <div
      class="ui grid"
    >
      <div
        class="three column row"
      >
        <div
          class="column"
        >
          <div
            class="field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="name"
                placeholder="Name"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="surname"
                placeholder="Surname"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="username"
                placeholder="Username"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="three column row"
      >
        <div
          class="column"
        >
          <div
            class="field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="email"
                placeholder="Email"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="password"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="birthdate"
                placeholder="Birthdate"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="three column row"
      >
        <div
          class="column"
        >
          <div
            class="field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="country"
                placeholder="Country"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="city"
                placeholder="City"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="address"
                placeholder="Address"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="one column row"
      >
        <div
          class="column"
        >
          <div
            class="field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="postal"
                placeholder="Postal"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="one column row"
      >
        <div
          class="center aligned column"
        >
          <button
            class="ui huge primary button"
          >
            Submit
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
